File: docker/Transformation-Engine/src/main/java/nhsd/fhir/converter/service/mapping/ConverterService.java
package nhsd.fhir.converter.service.mapping;

import ca.uhn.fhir.context.FhirVersionEnum;
import org.hl7.fhir.instance.model.api.IBaseResource;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.xml.sax.SAXException;

import javax.xml.parsers.ParserConfigurationException;
import java.io.IOException;

@Service
public class ConverterService {

    private final Converter converter;
    private final Transformer transformer;
    private final FhirParser fhirParser;

    public ConverterService(Converter converter, Transformer transformer, FhirParser fhirParser) {
        this.converter = converter;
        this.transformer = transformer;
        this.fhirParser = fhirParser;
    }

    public String convert(String resource, MediaType mediaType, FhirVersionEnum fhirVersion) throws ParserConfigurationException, IOException, SAXException {
        Class<? extends IBaseResource> resourceType = ResourceTypeFactory.createResourceType(resource, mediaType, fhirVersion);

        IBaseResource fhirResource = fhirParser.parse(resource, resourceType, mediaType);
        IBaseResource converted = converter.convert(fhirResource, fhirVersion);
        IBaseResource transformed = transformer.transform(converted);

        FhirVersionEnum outVersion = fhirVersion == FhirVersionEnum.R4 ? FhirVersionEnum.DSTU3 : FhirVersionEnum.R4;

        return fhirParser.encode(transformed, mediaType, outVersion);
    }
}

File: docker/Transformation-Engine/src/main/kotlin/net/nhsd/fhir/converter/service/ConverterService.kt
package net.nhsd.fhir.converter.service

import ca.uhn.fhir.context.FhirVersionEnum
import net.nhsd.fhir.converter.Converter
import net.nhsd.fhir.converter.FhirParser
import net.nhsd.fhir.converter.findResolvedAllergy
import net.nhsd.fhir.converter.getResourceType
import net.nhsd.fhir.converter.transformer.Transformer
import org.springframework.http.MediaType
import org.springframework.stereotype.Service

@Service
class ConverterService(
    private val fhirParser: FhirParser,
    private val careconnectTransformer: Transformer,
    private val converter: Converter,
) {

    fun convert(
        resource: String,
        inMediaType: MediaType,
        inFhirVersion: FhirVersionEnum,
        outMediaType: MediaType,
        outFhirVersion: FhirVersionEnum
    ): String {
        val resourceType = getResourceType(resource, inMediaType, inFhirVersion)
        val parsed = fhirParser.parse(resource, inMediaType, resourceType)

        // collect Ids for resolved allergies if doing dstu3 -> r4 conversion
        val resolvedAllergyIds = findResolvedAllergy(parsed, inFhirVersion, outFhirVersion)

        val converted = converter.convert(parsed, inFhirVersion, outFhirVersion)

        val transformed = careconnectTransformer.transform(parsed, converted)

        // handle resolved allergy

        return fhirParser.encode(transformed, outMediaType, outFhirVersion)
    }
}
